fix(table): updated expandable to animate visibility

mcoker opened the pull request:
fixes https://github.com/patternfly/patternfly/issues/7536

mcoker committed:
fix(table): updated expandable to animate visibility

patternfly-build commented:
Preview: https://patternfly-pr-7537.surge.sh

A11y report: https://patternfly-pr-7537-a11y.surge.sh

thatblindgeye reviewed (commented):
Tested this against the React PR that resolves the divider issue https://github.com/patternfly/patternfly-react/pull/11815 and it mostly looks good. The one thing I'm noticing is when you click to collapse a table row, the divider line appears as the expandable row is collapsing (causing the table to look like it has an empty row momentarily). I think it's really only noticeable at 25 or 10% animation speed, though.

My mac isn't currently letting my screen record so pictures instead:

![PatternFly table example showing how expandable animations causes the table to momentarily look like it has an empty table row as the expandable row is animating out.](https://github.com/user-attachments/assets/a62c4a41-df14-4dbb-b565-9e178efc91f3)

srambach commented:
> Tested this against the React PR that resolves the divider issue [patternfly/patternfly-react#11815](https://github.com/patternfly/patternfly-react/pull/11815) and it mostly looks good. The one thing I'm noticing is when you click to collapse a table row, the divider line appears as the expandable row is collapsing (causing the table to look like it has an empty row momentarily). I think it's really only noticeable at 25 or 10% animation speed, though.

Does it have to do with the `border-block-end: revert` on line 1051 and/or not everything is getting marked as expanded as it should?
There are 3 elements with `.pf-m-expanded` and if you toggle the `table__tr` that's not the expandable row, it creates an extra border

<img width="1943" alt="image" src="https://github.com/user-attachments/assets/36eea366-a9ea-424a-8cc1-99f6569a4075" />

srambach commented:
There's a problem with the [compound expansion example](https://patternfly-pr-7537.surge.sh/components/table/#compound-expansion-example) - looks like maybe changing them to visibility instead of display:none means they are taking space.

mcoker committed:
chore(table): fix hidden issue

thatblindgeye reviewed (approved):
Other than ^ lgtm, animations work nicely when pulled into React (with the fixes for the divider bug that got merged in react)

mcoker commented:
@srambach can you check again?

I'm assuming we should animate compound expandable, too? At least the initial click on a compound expandable item, but also possibly a fade between items once one is expanded when clicking on a different item. cc @andrew-ronaldson wdyt? Here's a link to compound expandable in react - https://www.patternfly.org/components/table/react/compound-expandable/

sg00dwin reviewed (approved):
LGTM

mcoker committed:
fix(table): make sure all expandable tables are wrapped in exp-row-content

mcoker committed:
fix(table): fix compound expandable examples

mcoker committed:
fix(table): remove comp exp control row expanded border per drewski

mcoker committed:
chore(table): move animations to opt-in

mcoker committed:
chore(motion): bump max-height for all components

srambach commented on src/patternfly/components/Table/examples/Table.md:
```diff
@@ -1344,8 +1362,10 @@ Note: To apply padding to `.pf-v6-c-table__expandable-row`, wrap the content in
 | `.pf-v6-c-table__control-row` | `.pf-v6-c-table__expandable > <tr>` | Modifies a compound expandable table control row. |
-| `.pf-m-expanded` | `<tbody>`, `.pf-v6-c-table__compound-expansion-toggle` > `.pf-v6-c-button` | Modifies a tbody with a row and an expandable row. |
 | `.pf-v6-c-table__compound-expansion-toggle` | `<td>` | Modifies a `<td>` on active/focus. |
+| `.pf-m-expanded` | `<tbody>`, `.pf-v6-c-table__compound-expansion-toggle` > `.pf-v6-c-button` | Modifies a tbody with a row and an expandable row. |
```
This wording is confusing - modifies a tbody that contains ... and an expanded toggle (?)

mcoker replied:
That was already there (I didn't add it) but I took a stab. It looks like the selectors were wrong, too.

srambach commented on src/patternfly/components/Table/examples/Table.md:
```diff
@@ -1344,8 +1362,10 @@ Note: To apply padding to `.pf-v6-c-table__expandable-row`, wrap the content in
-| `.pf-m-expanded` | `<tbody>`, `.pf-v6-c-table__compound-expansion-toggle` > `.pf-v6-c-button` | Modifies a tbody with a row and an expandable row. |
 | `.pf-v6-c-table__compound-expansion-toggle` | `<td>` | Modifies a `<td>` on active/focus. |
+| `.pf-m-expanded` | `<tbody>`, `.pf-v6-c-table__compound-expansion-toggle` > `.pf-v6-c-button` | Modifies a tbody with a row and an expandable row. |
+| `.pf-m-no-background` | `.pf-v6-c-table__expandable-row-content` | Modifies the expandable row content to have a transparent background. For in compound expandable when the parent expandable row has no padding with `.pf-m-no-padding`. |
```
```suggestion
| `.pf-m-no-background` | `.pf-v6-c-table__expandable-row-content` | Modifies the expandable row content to have a transparent background. For use in compound expandable when the parent expandable row has no padding with `.pf-m-no-padding`. |
```

srambach commented on src/patternfly/components/Table/table.scss:
```diff
@@ -1008,24 +1050,16 @@
-  }
-
-  &:not(.pf-m-expanded) {
-    display: none;
```
Does this need to remain or use another method to hide rows that aren't expanded?

mcoker replied:
Added back, missed that completely. That hides the non-expanded rows in a non-expandable table.

srambach reviewed (commented):
It looks like compound expandable shows all the non-expanded rows unless `.pf-m-animate-expand` is on the table. Are we keeping `display: none` in that case so we don't break core users?

Also, I think we need to add `.pf-m-animate-expand` in the docs at the appropriate spot.

mcoker committed:
chore(table): hide collapsed rows for non-exp table

mcoker committed:
chore(table): docs update, css fix

mcoker commented:
@srambach thanks! I should have caught the non-animated rows showing!

Updated and added new examples with the animated class for regular and compound expandable.

thatblindgeye reviewed (changes requested):
Repeat from Slack convo below, but a couple things I'm noticing when pulling these updates into React with updates made there:

- In Core we need to apply the `pf-m-animate-expand` class to the outer table element. Doing that will show an issue seen in React where when anything but the 1st compound expandable section is expanded, there's a border above the expandable content. I assume it's due to some `tr + tr` selector not taking into account this animate-expand modifier? The core compound expandable looks fine when updating the classes so the second compound expansion button+content is expanded, it's not until you add the missing `pf-m-animate-expand` to the outer table where you can reproduce the issue
- If content within the `pf-v6-c-table__expandable-row-content` element has a margin (as it does in React examples/demos for compound expansion), that margin continues to take up space when collapsed. Updating the React examples/demos to use the padding utility class instead of the margin one should resolve the issue, but just wondering if we want to handle a case of margin being used.

Not from Slack, but also examples will need to be updates so that `hidden` and `inert` are added to the collapsed rows.

mcoker committed:
fix(table): fix compound expandable border

mcoker committed:
fix(table): hide collapsed row overflow

mcoker committed:
fix(table): fix border issue

srambach commented:
I think now the top padding is missing on tabs that aren't the first one?  You can see in the inspector that `padding-block-start` is 0 if it's not immediately following a control row

https://github.com/user-attachments/assets/ce3c9e42-39d0-45db-b4be-8251885395fb

mcoker committed:
fix(table): fix expandable top padding

thatblindgeye reviewed (approved):
Looks good when linked to my React draft!

srambach reviewed (approved):
I'm good with a follow-on later to see if we can improve the bounce when switching from one compound expandable to another.

patternfly-build commented:
:tada: This PR is included in version 6.3.0-prerelease.28 :tada:

The release is available on:
- [GitHub release](https://github.com/patternfly/patternfly/releases/tag/prerelease-v6.3.0-prerelease.28)
- [npm package (@prerelease dist-tag)](https://www.npmjs.com/package/@patternfly/patternfly/v/6.3.0-prerelease.28)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
